samza-core/src/main/java/org/apache/samza/execution/StreamManager.java

```diff
@@ -75,4 +85,52 @@ public void createStreams(List<StreamSpec> streams) {
 
     return streamToPartitionCount;
   }
+
+  /**
+   * This is a best-effort approach to clear the internal streams from previous run, including intermediate streams,
+   * checkpoint stream and changelog streams.
+   * For batch processing, we always clean up the previous internal streams and create a new set for each run.
+   * @param prevConfig config of the previous run
+   */
+  public void clearStreamsFromPreviousRun(Config prevConfig) {
+    try {
+      ApplicationConfig appConfig = new ApplicationConfig(prevConfig);
+      LOGGER.info("run.id from previous run is {}", appConfig.getRunId());
+
+      StreamConfig streamConfig = new StreamConfig(prevConfig);
+
+      //Find all intermediate streams and clean up
+      Set<StreamSpec> intStreams = JavaConversions.asJavaCollection(streamConfig.getStreamIds()).stream()
+          .filter(streamConfig::getIsIntermediate)
+          .map(id -> new StreamSpec(id, streamConfig.getPhysicalName(id), streamConfig.getSystem(id)))
+          .collect(Collectors.toSet());
+      intStreams.forEach(stream -> {
+          LOGGER.info("Clear intermediate stream {} in system {}", stream.getPhysicalName(), stream.getSystemName());
+          sysAdmins.get(stream.getSystemName()).clearStream(stream);
+        });
+
+      //Find checkpoint stream and clean up
+      TaskConfig taskConfig = new TaskConfig(prevConfig);
+      String checkpointManagerFactoryClass = taskConfig.getCheckpointManagerFactory().getOrElse(defaultValue(null));
+      if (checkpointManagerFactoryClass != null) {
+        CheckpointManager checkpointManager = ((CheckpointManagerFactory) Util.getObj(checkpointManagerFactoryClass))
+            .getCheckpointManager(prevConfig, new MetricsRegistryMap());
+        checkpointManager.clearCheckpoints();
+      }
+
+      //Find changelog streams and remove them
```

Comment:
This is only sufficient if host affinity is disabled. Where do we enforce that?

Reply:
good point. I added the check inside ExecutionPlanner.validateConfig().